doc: add user guide on how to enable private key provider (#2533)
envoyproxy/gateway: site/content/en/latest/user/private-key-provider.md

---
title: "Using Envoy Private Key Provider for TLS termination"
---

> The Envoy Private Key Provider (https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-privatekeyprovider) allows implementing different provider which can provide the private key operation acceleration through the hardware or private key protection through the hardware.

Reviewer: ```suggestion
[The Envoy Private Key Provider](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-privatekeyprovider) allows implementing different provider which can provide the private key operation acceleration through the hardware or private key protection through the hardware.
```

> The Envoy Private Key Provider (https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-privatekeyprovider) allows implementing different provider which can provide the private key operation acceleration through the hardware or private key protection through the hardware.

Reviewer: can we simplify this line, so users can better understand why they would be interested in this feature

Author: ok, let me try. thanks

Today, there are two private key providers were implemented as Envoy contrib-extension:

>  CryptoMB in Envoy 1.20 release (https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto )

Reviewer: ```suggestion
 [CryptoMB in Envoy 1.20 release](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto )
```

> QAT in Envoy 1.24 release ( https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/private_key_providers/qat/v3alpha/qat.proto#extensions-private-key-providers-qat-v3alpha-qatprivatekeymethodconfig )

Reviewer: ```suggestion
[QAT in Envoy 1.24 release]( https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/private_key_providers/qat/v3alpha/qat.proto#extensions-private-key-providers-qat-v3alpha-qatprivatekeymethodconfig )
```

Both of them are used to accelerate the TLS handshake through the hardware capabilities.

This guide will walk through the steps required to configure TLS Terminate mode for TCP traffic and same time using the Envoy Private Key Provider to accelerate the TLS handshake.

## Prerequisites

> 

Reviewer: as a user reading this doc, would prefer if this section read as 
* how do I know if my HW has this capability or not
* how do I turn on the HW/SW capability

---

> title: "Using Envoy Private Key Provider for TLS termination"

Reviewer: ```suggestion
title: "Accelerating TLS Handshakes using Private Key Provider in Envoy"
```

---

> To allow the accelerate the private key operation through hardware or protect the private key through the hardware, the [Envoy Private Key Provider](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-privatekeyprovider) is added to Envoy.

Reviewer: ```suggestion
TLS operations can be accelerated or the private key can be protected using specilized hardware. This can be leveraged in Envoy using [Envoy Private Key Provider](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-privatekeyprovider).
```

> Today, there are two private key providers were implemented as Envoy contrib-extension:

Reviewer: ```suggestion
Today, there are two private key providers implemented in Envoy as contrib extensions:
```

- [QAT in Envoy 1.24 release](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/private_key_providers/qat/v3alpha/qat.proto#extensions-private-key-providers-qat-v3alpha-qatprivatekeymethodconfig)
- [CryptoMB in Envoy 1.20 release](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto )

Both of them are used to accelerate the TLS handshake through the hardware capabilities.

This guide will walk through the steps required to configure TLS Terminate mode for TCP traffic and same time using the Envoy Private Key Provider to accelerate the TLS handshake.

## Prerequisites

### For QAT

- Install Linux kernel 5.17 or similar

> - Ensure the node has QAT devices by checking the QAT physical function devices presented.

Reviewer: how can I confirm yes ?

Author: the command as below

  ```shell
  dmesg| grep IOMMU
  [    1.528237] DMAR: IOMMU enabled
  ```

- Enable virtual function devices for QAT device

  ```shell
  rmmod qat_4xxx
  modprobe qat_4xxx
  qat_device=$(lspci -D -d :4940 | awk '{print $1}')
  for i in $qat_device; do echo 16|sudo tee /sys/bus/pci/devices/$i/sriov_numvfs; done
  ```

- Enhance the container runtime memory lock limit

  ```shell
  mkdir /etc/systemd/system/containerd.service.d
  cat <<EOF >>/etc/systemd/system/containerd.service.d/memlock.conf
  [Service]
  LimitMEMLOCK=134217728
  EOF
  ```

  Restart the container runtime (for containerd, CRIO has similar concept)

  ```shell
  systemctl daemon-reload
  systemctl restart containerd
  ```

- Install [Intel® QAT Device Plugin for Kubernetes](https://github.com/intel/intel-device-plugins-for-kubernetes)

  ```shell
  kubectl apply -k 'https://github.com/intel/intel-device-plugins-for-kubernetes/deployments/qat_plugin?ref=main'
  ```

  Verification of the plugin deployment and detection of QAT hardware can be confirmed by examining the resource allocations on the nodes:

  ```shell
  kubectl get node -o yaml| grep qat.intel.com
  ```

### For CryptoMB:

> It required the node with 3rd generation Intel Xeon Scalable processor server processors, or later.

Reviewer: lets add the o/p from the commands so a user is certain their platform can support this feature

Author: ok, got it.

Author: I found I already added in the below

- For kubernetes Cluster, if not all nodes that support Intel® AVX-512 in Kubernetes cluster, you need to add some labels to divide these two kinds of nodes manually or using [NFD](https://github.com/kubernetes-sigs/node-feature-discovery).

  ```shell
  kubectl apply -k https://github.com/kubernetes-sigs/node-feature-discovery/deployment/overlays/default?ref=v0.15.1
  ```

- Checking the available nodes with required cpu instructions:
  - Check the node labels if using [NFD](https://github.com/kubernetes-sigs/node-feature-discovery):

    ```shell
    kubectl get nodes -l feature.node.kubernetes.io/cpu-cpuid.AVX512F,feature.node.kubernetes.io/cpu-cpuid.AVX512DQ,feature.node.kubernetes.io/cpu-cpuid.AVX512BW,feature.node.kubernetes.io/cpu-cpuid.AVX512VBMI2,feature.node.kubernetes.io/cpu-cpuid.AVX512IFMA
    ```

  - Check CPUIDS manually on the node if without using NFD:

    ```shell
    cat /proc/cpuinfo |grep avx512f|grep avx512dq|grep avx512bw|grep avx512_vbmi2|grep avx512ifma
    ```

## Installation

* Follow the steps from the [Quickstart Guide](../quickstart) to install Envoy Gateway.

> * Enables the EnvoyPatchPolicy feature.

Reviewer: ```suggestion
* Lets enable the EnvoyPatchPolicy feature, which will allow us to directly configure the Private Key Provider Envoy Filter, since Envoy Gateway does not directly expose this functionality.
```

EOF
```

### For CryptoMB

```shell
cat <<EOF | kubectl apply -f -
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: EnvoyPatchPolicy
metadata:
  name: key-provider-patch-policy
  namespace: default
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
    namespace: default
  type: JSONPatch
  jsonPatches:
    - type: "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret"
      name: default-example-cert
      operation:
        op: add
        path: "/tls_certificate/private_key_provider"
        value:
          provider_name: cryptomb
          typed_config:  
            "@type": "type.googleapis.com/envoy.extensions.private_key_providers.cryptomb.v3alpha.CryptoMbPrivateKeyMethodConfig"
            private_key:
              inline_string: |
                abcd
            poll_delay: 0.001s
          fallback: true
    - type: "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret"
      name: default-example-cert
      operation:
        op: copy
        from: "/tls_certificate/private_key"
        path: "/tls_certificate/private_key_provider/typed_config/private_key"
EOF
```

## Testing

> Follow the instructions in [TLS Termination for TCP](./tls-termination) to do the test.

Reviewer: can we create two listeners or gateway ? to test with and w/o key provider ? and highlight the differentiating factor ? e.g. time to complete the request in client or some way to show time taken to complete tls handshake in envoy

Author: ok, we should be able to do that.

Author: I added fortio command before enabling the envoy patch policy and after to benchmark the gateway simply. Then the people can compare the results to know the private key provider is working.

Today, there are two private key providers implemented in Envoy as contrib extensions:
- [QAT in Envoy 1.24 release](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/private_key_providers/qat/v3alpha/qat.proto#extensions-private-key-providers-qat-v3alpha-qatprivatekeymethodconfig)
- [CryptoMB in Envoy 1.20 release](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto )

Both of them are used to accelerate the TLS handshake through the hardware capabilities.

This guide will walk you through the steps required to configure TLS Termination mode for TCP traffic while also using the Envoy Private Key Provider to accelerate the TLS handshake.

## Prerequisites

### For QAT

- Install Linux kernel 5.17 or similar
- Ensure the node has QAT devices by checking the QAT physical function devices presented.

  ```shell

>   lspci -d 8086:4940

Reviewer 2: I suggest attaching the [documentation](https://intel.github.io/quickassist/qatlib/requirements.html#qat2-0-qatlib-supported-devices) of the dependency. Since different QAT generations have different device IDs.

Author: thanks, I will paste the command in the doc and reference the doc

  ```

- Enable IOMMU from BIOS
- Enable IOMMU for Linux kernel

  ```shell
  cat /etc/default/grub:
  GRUB_CMDLINE_LINUX="intel_iommu=on vfio-pci.ids=8086:4941"
  update-grub
  reboot
  ````

  Once the system is rebooted, check if the IOMMU has been enabled via the following command:

  ```shell
  dmesg| grep IOMMU
  [    1.528237] DMAR: IOMMU enabled
  ```

- Enable virtual function devices for QAT device

  ```shell
  rmmod qat_4xxx
  modprobe qat_4xxx
  qat_device=$(lspci -D -d :4940 | awk '{print $1}')
  for i in $qat_device; do echo 16|sudo tee /sys/bus/pci/devices/$i/sriov_numvfs; done

>   ```

Reviewer 2: Missing `modprobe vfio_pci` at the beginning and `chmod a+rw /dev/vfio/*` at the end?

Author: thanks！

> - Enhance the container runtime memory lock limit

Reviewer 2: Do you mean "increase" here? And it would be better to mention the configuration file depends on the container runtime.

Author: good point
